osu.Game/Modes/Ruleset.cs

```diff
@@ -0,0 +1,36 @@
+﻿//Copyright (c) 2007-2016 ppy Pty Ltd <contact@ppy.sh>.
+//Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using System.Collections.Generic;
+using osu.Game.Modes.Objects;
+using osu.Game.Modes.UI;
+using System.Reflection;
+using osu.Framework.Extensions;
+using System;
+using System.Linq;
+
+namespace osu.Game.Modes
+{
+    public abstract class Ruleset
+    {
+        private static List<Type> availableRulesets = new List<Type>();

```

Comment:
Make this a `ConcurrentSet` such that public static methods that write data can not mess things up (either by threading or by adding a ruleset twice).